### PR TITLE
Add Protection against mention flooding

### PR DESCRIPTION
**Pull request by MTRNord**
This basically tries to solve the current way of spam.

This is an adapted wordfilter protection. At the time of writing, this is not tested anywhere.

It only matches mxid based mentions currently. Not other mentions per default push rules.

## TODO

- [x] Finish the warning branch
- [x] Test it (I have a docker of this branch at `coreharbor.kubernetes.midnightthoughts.space/matrixdotorg/mjolnir:mentionflood-protection` )

image
image

Signed-off-by: Marcel Radzio <mtrnord[at]nordgedanken.dev>

### CODE REVIEW

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:53]: This should move into `settings`, so that we can change it dynamically without having to restart Mjölnir.

See BasicFlooding.ts for an example.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:64]: We should sanitize this input with `htmlEscape` (or is it `escapeHtml`, I forgot?)

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:39]: Please use `Map`s. Less chances of a bad surprise, if e.g. the `username` is something like `__proto`.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:85]: Here too, sanitization.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:99]: Shouldn't this be `matchAll`?

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:43]: Why is the regex so complex? Wouldn't it be sufficient (and faster) to just check something along the lines of `\b@[^\b]+:`?

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:104]: Sanitization.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:101]: This also should move to `settings`. Also, it would be much better if we could have a choice between `kick`, `ban`, `warn` (to do nothing).

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:108]: Also a setting. Probably another one called `redact`, expecting a `bool`.

**Review by Yoric — CHANGES_REQUESTED** [review]
Thanks a lot for this PR!

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts:99]: As I do the g flag above, I count the amount of full matches here. The all part should already happen using the g flag. I will double-check this later on my mjolnir.

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts:43]: Mostly because my regex knowledge isnt the best and I went by spec to cover all the cases it mentions for a mxid.

Comment (Yoric) [reply src/protections/MentionFlood.ts:54]: I've just double-checked, my suggestion doesn't work.

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts]: Hm where should the warning actually happen? Warning the moderator or warning the user?

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts:54]: Hm actually mine I think probably also fails on historic IDs 🤔 I only check for the new mxid syntax in this.

Comment (MTRNord) [reply src/protections/MentionFlood.ts]: Mine in theory should now work for historic ones too. I think that can be simplified, as the current localparts are a subset of the historic ones.

**Comment by MTRNord** [comment]
Seems like my vscode likes to reformat stuff differently somehow :/ I can revert the unrelated formatting changes if needed.

File: src/protections/MentionFlood.ts
Comment (MTRNord) [inline src/protections/MentionFlood.ts:1]: (Making this a new thread, so it can easily get replied on)

@deepbluev7 in a room mentioned, it might make sense to also match mentions based on the default pushrule behaviour. So for me, "mtrnord" would also mention me.

Should that go into this PR, or should that be in a follow up PR? (Asking as I think it makes sense to have it, but the current wave seems to mainly use mxids?)

File: src/protections/MentionFlood.ts
Comment (Yoric) [reply src/protections/MentionFlood.ts]: We don't care all that much about old users, mostly about new ones.

File: src/protections/MentionFlood.ts
Comment (Yoric) [reply src/protections/MentionFlood.ts]: Warning the moderator.

File: src/protections/MentionFlood.ts
Comment (Yoric) [reply src/protections/MentionFlood.ts:1]: Let's start with a simple PR, then expand in a followup.

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts]: It seems like there may actually be a bug somewhere in the check as I cant get it to work

Comment (MTRNord) [reply src/protections/MentionFlood.ts]: Bug was escaping the regex. It works now

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts]: This seems to work now as wanted. The actual issue with it was that I didnt escape the backslashes as required

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts]: > We don't care all that much about old users, mostly about new ones.

Ok fair. Should I get rid of the historic user id part in the regex?

Comment (Yoric) [reply src/protections/MentionFlood.ts]: I think so.

**Comment by Yoric** [comment]
Oops, looks like CI was waiting for approval before launching.

Let's see how it goes.

File: src/protections/MentionFlood.ts
Comment (jesopo) [inline src/protections/MentionFlood.ts:91]: do any clients opt for showing `body` even if `formatted_body` is present? if so we'll want to check both

File: src/protections/MentionFlood.ts
Comment (jesopo) [inline src/protections/MentionFlood.ts:47]: I'd create a `OptionListProtectionSetting` for use as `OptionListProtectionSetting(["true", "false"])`, which would subclass of `StringListProtectionSetting` but have a set of strings it thinks is valid

Comment (jesopo) [reply src/protections/MentionFlood.ts:47]: perhaps additional to this, in a later PR, we can somehow support `AbstractProtectionSetting.validate()` returning a human readable error

**Review by jesopo — COMMENTED** [review]
big thank for the PR

File: src/protections/MentionFlood.ts
Comment (ShadowJonathan) [reply src/protections/MentionFlood.ts:91]: Clients show formatted_body if preferable, but mentions (currently) are only checked per the body, the formatted body does not get inspected for mentions/pings

Comment (MTRNord) [reply src/protections/MentionFlood.ts:91]: The order chosen here is mostly as only the formatted body of ele Web includes the mxid. The body only has the display names. And the code explicitly checks for mxids

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts]: I adapted this to be not a list returned but a thing with only one value as I think that makes sense here. If not please explain this to me in detail :) I may be missing something here.

**Comment by cremesk** [comment]
can we merge it please? i really need this protection :D

File: src/protections/MentionFlood.ts
Comment (jesopo) [reply src/protections/MentionFlood.ts:89]: that makes sense.

Comment (jesopo) [reply src/protections/MentionFlood.ts:89]: this is definitely the case for pill-style mentions (I've just checked) but otherwise I don't think so. this will probably be good enough for now though.

Comment (MTRNord) [reply src/protections/MentionFlood.ts:89]: The alternative would be just to check both and go with the higher number I guess

**Review by jesopo — APPROVED** [review]

**Comment by Yoric** [comment]
@MTRNord Just back from trip, I'll try and review this today or tomorrow.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:49]: FWIW, we now have `mjolnir.roomJoins.getUserJoin` that implements this feature in such a manner that all protections can use it.

Ideally, I'd prefer if you could use this instead of `justJoined`, as I'm a bit afraid it's going to grow very large for rooms such as MatrixHQ, so I'd like to avoid having two copies of this in memory. Do you feel that you have time to work on this? Otherwise, this can happen in a followup issue.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:71]: Nit: Let's always put `{}` on a `if`, this can avoid accidents.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:97]: Nit:
```js
const now = Date.now();
```
will save allocating an object.

File: src/protections/MentionFlood.ts
Comment (Yoric) [inline src/protections/MentionFlood.ts:45]: ...in particular, `redact` is not self-explanatory.

File: src/protections/ProtectionSettings.ts
Comment (Yoric) [inline src/protections/ProtectionSettings.ts:228]: Nit: Missing newline.

**Review by Yoric — COMMENTED** [review]
PR looks good, could you add a test?

**Comment by Yoric** [comment]
@MTRNord I realize I wasn't clear: when I write test, I mean a new integration test, see e.g. `test/integration`.

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts:66]: Sounds good to me :)

File: src/protections/MentionFlood.ts
Comment (MTRNord) [reply src/protections/MentionFlood.ts:75]: This one was removed completely, so this if is not anymore needed :)

**Comment by Yoric** [comment]
@MTRNord What's the status of this PR?

**Comment by MTRNord** [comment]
> @MTRNord What's the status of this PR?

Sorry, I nearly have the tests done, but there is some small thing I still need to do. Just didn't get to it yet due to other work related tasks that took higher priority. Want to do the rest of the code on the weekend.

**Comment by Yoric** [comment]
Don't hesitate to yell if you need help!

File: test/integration/protections/MentionFloodTest.ts
Comment (Gnuxie) [inline test/integration/protections/MentionFloodTest.ts:93]: nitpick but `userA` doesn't tell us what this use is doing. they could unrelated to moderation of the room, they could just be a participant, they could even be a user who is going to be spamming, who is `userA`?

**Review by Gnuxie — COMMENTED** [review]
Made some observations about the test, you don't need to act on them but they're just things that would keep them cleaner this or next time

**Comment by Yoric** [comment]
Sorry for the delay, I'm just back from vacation, I'll try and review this today.

File: test/integration/protections/protectionUtils.ts
Comment (Yoric) [inline test/integration/protections/protectionUtils.ts:85]: Nit: Missing newline.

**Review by Yoric — CHANGES_REQUESTED** [review]
Quite good!
You'll need a few more changes, but we're getting there.

**Pull request closed without merging**
